google/ingester.go

```diff
@@ -0,0 +1,327 @@
+					machinteTypePrices[fmt.Sprintf("%s.%s", mf, group)] = pwp.Price
+				}
+				if ing.ingestionFilter(pwp) {
+					results <- pwp
+				}
+			}
+		}
+
+		for mt := range ing.fetchMachineTypes(ctx) {
+			mf := strings.ToLower(strings.Split(mt.Name, "-")[0])
+			_, ok := machineFamilies[mf]
+			if !ok {
+				continue
+			}
+
+			prod := &product.Product{
+				Provider: ProviderName,
+				SKU:      fmt.Sprintf("machinte-types-%d", mt.Id),
+				Service:  "Compute Engine",
+				Family:   "Compute",
+				Location: ing.region,
+				Attributes: map[string]string{
+					"machine_type":   mt.Name,
+					"group":          "MachineType",
+					"cpu":            strconv.Itoa(int(mt.GuestCpus)),
+					"ram":            strconv.Itoa(int(mt.MemoryMb)),
+					"kind":           mt.Kind,
+					"machine_family": mf,
+				},
+			}
+
+			rp, ok := machinteTypePrices[fmt.Sprintf("%s.%s", mf, ram)]
+			if !ok {
+				continue
+			}
+
+			cp, ok := machinteTypePrices[fmt.Sprintf("%s.%s", mf, cpu)]
+			if !ok {
+				continue
+			}
+
+			// The mt.MemboryMb is in MB and the rp.Unit is on GiBy
+			// so we have to convert it to the same unit
+			rp.Value = rp.Value.Mul(decimal.NewFromInt(mt.MemoryMb / 1_000))
+			cp.Value = cp.Value.Mul(decimal.NewFromInt(mt.GuestCpus))
+
+			// The Unit for the RAM is in GiBy.h, but for the addition
+			// we need the same unit which in this case would be h
+			if rp.Unit == "GiBy.h" {
+				rp.Unit = "h"
+			} else {
+				// If we cannot check the unit we skip it
+				continue
+			}
+			err := rp.Add(cp)
+			if err != nil {
+				ing.err = err
+				return
+			}
+
+			pwp := &price.WithProduct{
+				// r has the end result of the Sum of the 2 prices
+				Price:   rp,
+				Product: prod,
+			}
+			if ing.ingestionFilter(pwp) {
+				results <- pwp
+			}
+		}
+	}()
+
+	return results
+}
+
+// fetchSkus returns a channel where the Sku are beeing sent and pulls all the
+// data from the Skus paginating through all the API
+func (ing *Ingester) fetchSkus(ctx context.Context) <-chan *cloudbilling.Sku {
+	results := make(chan *cloudbilling.Sku, 100)
+
+	go func() {
+		defer close(results)
+		// Docs: https://cloud.google.com/billing/v1/how-tos/catalog-api#getting_the_list_of_skus_for_a_service
+		err := cloudbilling.NewServicesSkusService(ing.billing).List(fmt.Sprintf("services/%s", ing.service)).Pages(ctx, func(l *cloudbilling.ListSkusResponse) error {
+			for _, sku := range l.Skus {
+				results <- sku
+			}
+			return nil
+		})
+		if err != nil {
+			ing.err = err
+		}
+	}()
+
+	return results
+}
+
+// fetchMachineTypes returns a channel where the machine types are beeing sent and pulls all the
+// data from the machine types paginating through all the API
+func (ing *Ingester) fetchMachineTypes(ctx context.Context) <-chan *compute.MachineType {
+	results := make(chan *compute.MachineType, 100)
+
+	go func() {
+		defer close(results)
+		// Docs: https://cloud.google.com/compute/docs/reference/rest/v1/machineTypes/list
```

Comment:
According to the documentation :
```
The maximum number of results per page that should be returned. If the number of available results is larger than maxResults, Compute Engine returns a nextPageToken that can be used to get the next page of results in subsequent list requests. Acceptable values are 0 to 500, inclusive. (Default: 500)
```

I'm not sure how many types of Machines there is, but couldn't you end up in a deadlock situation in case you have more than 100?

Reply:
No, the channel size means how much buffer it can have until it's blocked and you cannot push more data until someone reads from it